Sources/SourceKitLSP/Swift/CodeActions/ApplyDeMorganLaw.swift

```diff
@@ -0,0 +1,730 @@
+  /// ~(~a | ~b) /* becomes */ (a & b)
+  /// ```
+  struct RootExpr {
+    private var expr: Expr
+
+    /// Attempts to fold `unstructuredExpr` into a structured expression using `SwiftOperators` and stores the result.
+    init?(unstructuredExpr: ExprSyntax) {
+      if let structuredExpr = OperatorTable.standardOperators.foldAll(
+        unstructuredExpr,
+        errorHandler: { _ in
+        }
+      ).as(ExprSyntax.self) {
+        self.expr = Expr(expr: structuredExpr)
+      } else {
+        return nil
+      }
+    }
+
+    /// Returns the complement of the expression stored, if any.
+    func complement() -> ExprSyntax? {
+      self.complementOfNegation() ?? self.complementOfPropositions()
+    }
+
+    /// Returns the complement of the expression stored by treating it as a negation, if any.
+    private func complementOfNegation() -> ExprSyntax? {
+      if let prefixExpr = self.expr.as(PrefixOperatorExprSyntax.self),
+        let negation = Negation(prefixExpr),
+        let complement = Expr(expr: negation.expr).negatedPropositions(exprType: negation.exprType)
+      {
+        // TODO: remove parentheses?
+        self.expr.map(negatedExpr: complement.expr)
+      } else {
+        nil
+      }
+    }
+
+    /// Returns the complement of the expression stored by treating it as propositions, if any.
+    private func complementOfPropositions() -> ExprSyntax? {
+      if let infixExpr = self.expr.as(InfixOperatorExprSyntax.self),
+        let propositions = NegatedPropositions(infixExpr)
+      {
+        self.expr.map(
+          negation: Negation(
+            exprType: propositions.operator.exprType,
+            expr: propositions.expr
+          )
+        ).expr
+      } else {
+        nil
+      }
+    }
+  }
+
+  /// The type of expression, either bitwise or boolean.
+  private enum ExprType {
+    case bitwise
+    case boolean
+
+    init?(operator: TokenSyntax) {
+      switch `operator`.tokenKind {
+      case .prefixOperator("~"):
+        self = .bitwise
+      case .prefixOperator("!"):
+        self = .boolean
+      default:
+        return nil
+      }
+    }
+
+    var negationPrefix: TokenSyntax {
+      switch self {
+      case .bitwise:
+        .prefixOperator("~")
+      case .boolean:
+        .prefixOperator("!")
+      }
+    }
+  }
+
+  /// An expression wrapper abstracting away any presence of extraneous parentheses in the given expression, exposing
+  /// only the expression of interest for easier manipulation.
+  private enum Expr {
+    /// A single means a tuple containing only one element, implying that there exists at least one pair of extraneous
+    /// parentheses enclosing the expression of interest.
+    ///
+    /// - Parameters:
+    ///   - desinglified: The first non-single descendent expression. The expression of interest.
+    case single(TupleExprSyntax, desinglified: ExprSyntax)
+    /// A non-single expression at its root. The expression of interest.
+    case other(ExprSyntax)
```

Comment:
Do we really have to have an enum with these two cases? Wouldn’t it be easier to just operate on `ExprSyntax` and have a property that looks through all parentheses? Something like 

```swift
extension ExprSyntax {
  // Equivalent of the `desinglified` associated value in `single`
  var lookingThroughParentheses: ExprSyntax {
    if let tuple = self.as(TupleExprSyntax.self), 
      let underlyingExpr = tuple.singleUnlabledExpression?.lookingThroughParentheses else {
      return underlyingExpr
    }
    return self
  }

  // Equivalent of checking for the `single` case.
  var isParenthesizedExpression: Bool {
    return self.as(TupleExprSyntax.self)?.isParentheses ?? false
  }
}
```

Reply:
During this rewrite I've found explicit nominal types tremendously helpful for my thought process because I've been confused for so many times by conflating SwiftSyntax types with domain specific concepts. 

I'm open to remove these aids for thought at this stage ;)

Reply:
While reading through the code, they confused me more than they helped me. I usually thought: Why can’t we just return the rewritten expression? Also, the `map` functions taking both a syntax node and one of the types you defined confused me a bit because they seemed to reason about to very similar but only related, not equivalent types simultaneously.

You did spend more time thinking about this than I did and I might be missing something, but if you could find the energy to write an implementation without the intermediate types that you define, I would very much like to see that. Maybe it will convince me that the current design is the correct one and we find the correct pieces of documentation that I would have needed to reason about it.

Reply:
> While reading through the code, they confused me more than they helped me. I usually thought: Why can’t we just return the rewritten expression? Also, the `map` functions taking both a syntax node and one of the types you defined confused me a bit because they seemed to reason about to very similar but only related, not equivalent types simultaneously.
> 
It's a fair observation :) I'd leaned much towards adopting the terminology of the de Morgan's law when rewriting this code action because, for me, it'd provided a better conceptual mapping to the business logic by abstracting away the minute details of SwiftSyntax. I've struggled with and bikeshedded on naming everything, the picture didn't become clearer after I did a complete divorce from SwiftSyntax parlance.

> You did spend more time thinking about this than I did and I might be missing something, but if you could find the energy to write an implementation without the intermediate types that you define, I would very much like to see that. Maybe it will convince me that the current design is the correct one and we find the correct pieces of documentation that I would have needed to reason about it.

Sure, I can still expend some time doing a minor rewrite before the start of the new school term ;)